packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -3122,18 +3107,80 @@ export class StripeHelper extends StripeHelperBase {
 
     const productPaymentCycleOld = this.stripePlanToPaymentCycle(planOld);
 
+    const sameBillingCycle =
+      productPaymentCycleOld === baseDetails.productPaymentCycleNew;
+
+    // get next invoice details
+    const nextInvoice = upcomingInvoiceWithInvoiceItem
+      ? upcomingInvoiceWithInvoiceItem
+      : await this.previewInvoiceBySubscriptionId({
+          subscriptionId: subscription.id,
+        });
+
+    const { total: nextInvoiceTotal, currency: nextInvoiceCurrency } =
+      nextInvoice;
+
+    // https://stripe.com/docs/api/invoices/object#invoice_object-lines
+    // as per the Stripe docs, one of the individual line items includes
+    // new plan pricing and tax information without additional fees
+    const upcomingInvoiceLineItem =
+      nextInvoice.lines && nextInvoice.lines.data
+        ? nextInvoice.lines.data.find((line) => !line.proration)
+        : undefined;
+
+    // get new list price total including tax(es)
+    const newListPriceTotalWithoutTax =
+      upcomingInvoiceLineItem && upcomingInvoiceLineItem.tax_amounts
+        ? upcomingInvoiceLineItem.amount
+        : undefined;
+
+    const newListPriceTotalTaxes =
+      upcomingInvoiceLineItem && upcomingInvoiceLineItem.tax_amounts
+        ? upcomingInvoiceLineItem.tax_amounts.reduce(
+            (acc, tax) => acc + tax.amount,
+            0
+          )
+        : undefined;
+
+    const newListPriceTotal =
+      newListPriceTotalWithoutTax && newListPriceTotalTaxes
+        ? newListPriceTotalWithoutTax + newListPriceTotalTaxes
+        : undefined;
+
+    // if same billing cycle and line item exists, add amount and tax for new total
+    // else return the next invoice total
+    const newTotalInCents =
```

Comment:
Just to keep the PR up to date, as discussed in Slack, there seems to be an issue with this value when there's a discount applied, and for upgrades with plans that have inclusive tax.

Reply:
As per our Slack discussion, active discounts continue on upgrades of same billing cycle but not different billing cycle. This was not initially known but is expected behaviour. As a result, I implemented the fix here and in the upcoming epic or possibly later, we will have to make additional updates to include discount in subscription upgrade pricing and invoicing in other areas on the site (e.g. `PlanUpgradeDetails`, `PaymentConfirmation`, and `SubscriptionManagement`). I also resolved the inclusive tax issues and went ahead and also accounted for upgrading from inclusive to exclusive tax and vice versa.

**NOTE:** @StaberindeZA As there were over 16 different scenarios (see _Other information_ above), there is a chance one was missed or there could still be some edge cases that were unaccounted for. Here's crossing fingers we might finally have a fix in that tackles them.

Reply:
Totally aligned! Normally I would have added individual tests and improved code quality, but as mentioned in another comment, nearly all of these conditionals will be removed in favour of the else statement (currently different billing cycle > `always_invoice`) when FXA-5935 is done (estimated time of completion is next sprint and clean up two sprints from now).

FXA-6691, which this ticket in part blocks, ignores all this code under a flag (essentially the suggested if/else will be unnecessary). FXA-7796 will remove it entirely. Seeing as the latter is in progress and due for completion this sprint, the additional work will only add to more engineering effort and technical debt, as it's more code to be removed.

Note: this is a special case only because we're doing it in tandem with the epic to immediately invoice on upgrade.